File: src/leetcode/129.求根到叶子节点数字之和.go
package leetcode

/*
 * @lc app=leetcode.cn id=129 lang=golang
 *
 * [129] 求根到叶子节点数字之和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
	return helper(root, 0)
}

func helper(root *TreeNode, cur int) int {
	if root == nil {
		return 0 // 当前非叶子节点, 不计算
	}

	next := cur*10 + root.Val
	if root.Left == nil && root.Right == nil {
		return next // 当前为叶子节点, 计算
	}

	l := helper(root.Left, next)
	r := helper(root.Right, next)
	return l + r
}

// @lc code=end

File: src/leetcode/129.求根到叶子节点数字之和.php
<?php
require_once 'ds.php';

use PHPUnit\Framework\TestCase;

/*
 * @lc app=leetcode.cn id=129 lang=php
 *
 * [129] 求根到叶子节点数字之和
 */

// @lc code=start

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution
{

    /**
     * @param TreeNode $root
     * @return Integer
     */
    function sumNumbers(TreeNode $root)
    {
        return helper($root, 0);
    }
}

/**
 * @param TreeNode $root
 * @param int $cur
 * @return int
 */
function helper(TreeNode $root, int $cur)
{
    if (!$root) return $cur;
    $next = $cur * 10 + $root->val;
    if (!$root->left && !$root->right) return $next;

    $l = helper($root->left, $next);
    $r = helper($root->right, $next);
    return $l + $r;
}
// @lc code=end

